src/main/java/sirius/biz/scripting/ScriptBasedCustomEventDispatcher.java

```diff
@@ -0,0 +1,111 @@
+        handlers.put(eventType.getName(), handler);
+        active = true;
+    }
+
+    @Override
+    public <T, E extends TypedCustomEvent<T>> void registerTypedHandler(Class<E> eventType,
+                                                                        Class<T> type,
+                                                                        Callback<E> handler) {
+        handlers.put(buildTypedEventHandlerName(eventType, type), handler);
+        active = true;
+    }
+
+    private String buildTypedEventHandlerName(Class<?> eventType, Class<?> type) {
+        return eventType.getName() + "::" + type.getName();
+    }
+
+    @Override
+    public boolean isActive() {
+        return active;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void handleEvent(CustomEvent event) {
+        Callback<CustomEvent> handler = (Callback<CustomEvent>) handlers.get(determineEventKey(event));
+        if (handler == null) {
+            return;
+        }
+
+        Watch watch = Watch.start();
+        try {
+            handler.invoke(event);
+            event.success = true;
+        } catch (HandledException handledException) {
+            handleEventHandlerException(handledException);
+            event.failed = true;
+            event.error = handledException;
+        } catch (Exception e) {
+            HandledException handledException = Exceptions.handle(Scripting.LOG, e);
+            handleEventHandlerException(handledException);
+            event.failed = true;
+            event.error = handledException;
+        }
+        TaskContext.get().addTiming(NLS.get("CustomEventHandler.customEvents"), watch.elapsedMillis());
+    }
+
+    private String determineEventKey(CustomEvent event) {
+        if (event instanceof TypedCustomEvent<?> typedEvent) {
+            return buildTypedEventHandlerName(typedEvent.getClass(), typedEvent.getType());
+        } else {
+            return event.getClass().getName();
+        }
+    }
+
+    private void handleEventHandlerException(HandledException handledException) {
+        if (processes.fetchCurrentProcess().isPresent()) {
+            handleExceptionInProcess(handledException);
+        } else {
+            processes.executeInStandbyProcessForCurrentTenant("custom-event-handler",
+                                                              () -> NLS.get("CustomEventHandler.customEvents"),
+                                                              ignored -> handleExceptionInProcess(handledException));
+        }
+    }
+
+    private void handleExceptionInProcess(HandledException handledException) {
+        TaskContext.get()
+                   .log(NLS.fmtr("CustomEventHandler.message")
+                           .set("system", TaskContext.get().getSystemString())
+                           .set("message", handledException.getMessage())
+                           .format());
```

Comment:
**Optional:** Arent there logging capabilities where a NLS key and parameters can be supplied directly so the message is not stored translated but only translated in the correct display language when viewed?

Reply:
This API is not exposed via TaskContext and only available in Processes. While a Process will most probably be around, we want to decouple the two frameworks and rely on the more generic API of TaskContext here...